File: shared/video-protocol/ts/src/packets.ts
import {
  FRAME_START_MARKER,
  CHUNK_MARKER,
  CHUNK_PAYLOAD_SIZE,
  detectFrameStartFormat,
} from "./protocol";

// MARK: - FrameStartPacket

export interface FrameStartPacket {
  frameId:    bigint;
  frameSize:  bigint;
  chunkCount: bigint;
}

/**
 * Parse a FRAME_START UDP datagram.
 * Handles both 64-bit (primary, Raspberry Pi 64-bit) and 32-bit (fallback) wire formats.
 * Returns null for unrecognised data.
 */
export function parseFrameStart(buf: Buffer): FrameStartPacket | null {
  if (!buf.subarray(0, FRAME_START_MARKER.length).equals(FRAME_START_MARKER)) {
    return null;
  }

  const fmt = detectFrameStartFormat(buf);
  const body = buf.subarray(FRAME_START_MARKER.length);

  if (fmt === "uint64") {
    return {
      frameId:    body.readBigUInt64LE(0),
      frameSize:  body.readBigUInt64LE(8),
      chunkCount: body.readBigUInt64LE(16),
    };
  }

  if (fmt === "uint32") {
    return {
      frameId:    BigInt(body.readUInt32LE(0)),
      frameSize:  BigInt(body.readUInt32LE(4)),
      chunkCount: BigInt(body.readUInt32LE(8)),
    };
  }

  return null;
}

// MARK: - ChunkPacket

export interface ChunkPacket {
  frameId:    bigint;
  chunkIndex: bigint;
  payload:    Buffer;
}

/**
 * Parse a CHUNK UDP datagram.
 * Handles both 64-bit (primary) and 32-bit (fallback) wire formats.
 */
export function parseChunk(buf: Buffer): ChunkPacket | null {
  if (!buf.subarray(0, CHUNK_MARKER.length).equals(CHUNK_MARKER)) {
    return null;
  }

  const markerLen = CHUNK_MARKER.length;  // 5

  // Try 64-bit first (header = 5 + 16 = 21 bytes)
  if (buf.length >= 21) {
    return {
      frameId:    buf.readBigUInt64LE(markerLen),
      chunkIndex: buf.readBigUInt64LE(markerLen + 8),
      payload:    buf.subarray(21),
    };
  }

  // 32-bit fallback (header = 5 + 8 = 13 bytes)
  if (buf.length >= 13) {
    return {
      frameId:    BigInt(buf.readUInt32LE(markerLen)),
      chunkIndex: BigInt(buf.readUInt32LE(markerLen + 4)),
      payload:    buf.subarray(13),
    };
  }

  return null;
}

// MARK: - FrameAssembler

interface PendingFrame {
  buffer:          Buffer;
  expectedChunks:  number;
  receivedIndices: Set<number>;
}

/**
 * Reassembles UDP chunks into complete video frames.
 */
export class FrameAssembler {
  private pending = new Map<bigint, PendingFrame>();

  handleFrameStart(packet: FrameStartPacket): void {
    this.pending.set(packet.frameId, {
      buffer:          Buffer.alloc(Number(packet.frameSize)),
      expectedChunks:  Number(packet.chunkCount),
      receivedIndices: new Set(),
    });
  }

  /**
   * Returns the complete frame Buffer once all chunks arrive, otherwise null.
   */
  handleChunk(packet: ChunkPacket): Buffer | null {
    const frame = this.pending.get(packet.frameId);
    if (!frame) return null;

    const idx = Number(packet.chunkIndex);
    if (frame.receivedIndices.has(idx)) return null;

    const offset = idx * CHUNK_PAYLOAD_SIZE;
    if (offset >= frame.buffer.length) return null;

    packet.payload.copy(frame.buffer, offset, 0, frame.buffer.length - offset);
    frame.receivedIndices.add(idx);

    if (frame.receivedIndices.size >= frame.expectedChunks) {
      const complete = frame.buffer;
      this.pending.delete(packet.frameId);
      return complete;
    }

    return null;
  }

  /**
   * Drop stale partial frames, keeping only the newest `count` frame IDs.
   */
  pruneStale(keepLast = 30): void {
    if (this.pending.size <= keepLast) return;
    const sorted = [...this.pending.keys()].sort((a, b) => (a < b ? -1 : 1));
    sorted.slice(0, this.pending.size - keepLast).forEach((k) => this.pending.delete(k));
  }
}

File: shared/video-protocol/swift/Sources/VideoProtocol/Packets.swift
import Foundation

// MARK: - FrameStartPacket

/// Parsed FRAME_START packet.
///
/// Wire layout (64-bit / primary):
///   "FRAME_START" (11B) | frame_id (UInt64 LE) | frame_size (UInt64 LE) | chunk_count (UInt64 LE)
///   Total: 35 bytes
///
/// Wire layout (32-bit / fallback):
///   "FRAME_START" (11B) | frame_id (UInt32 LE) | frame_size (UInt32 LE) | chunk_count (UInt32 LE)
///   Total: 23 bytes
public struct FrameStartPacket {
    public let frameID:     UInt64
    public let frameSize:   UInt64
    public let chunkCount:  UInt64

    /// Parse from a raw UDP datagram. Returns nil for unrecognised data.
    public static func parse(from data: Data) -> FrameStartPacket? {
        let marker = VideoProtocol.frameStartMarker
        guard data.starts(with: marker) else { return nil }
        let body = data.dropFirst(marker.count)

        if data.count == 35 {            // 64-bit format (Raspberry Pi 64-bit, primary)
            return body.withUnsafeBytes { ptr -> FrameStartPacket? in
                guard ptr.count >= 24 else { return nil }
                let frameID    = UInt64(littleEndian: ptr.loadUnaligned(fromByteOffset:  0, as: UInt64.self))
                let frameSize  = UInt64(littleEndian: ptr.loadUnaligned(fromByteOffset:  8, as: UInt64.self))
                let chunkCount = UInt64(littleEndian: ptr.loadUnaligned(fromByteOffset: 16, as: UInt64.self))
                return FrameStartPacket(frameID: frameID, frameSize: frameSize, chunkCount: chunkCount)
            }
        } else if data.count >= 23 {    // 32-bit fallback
            return body.withUnsafeBytes { ptr -> FrameStartPacket? in
                guard ptr.count >= 12 else { return nil }
                let frameID    = UInt64(UInt32(littleEndian: ptr.loadUnaligned(fromByteOffset: 0, as: UInt32.self)))
                let frameSize  = UInt64(UInt32(littleEndian: ptr.loadUnaligned(fromByteOffset: 4, as: UInt32.self)))
                let chunkCount = UInt64(UInt32(littleEndian: ptr.loadUnaligned(fromByteOffset: 8, as: UInt32.self)))
                return FrameStartPacket(frameID: frameID, frameSize: frameSize, chunkCount: chunkCount)
            }
        }

        return nil
    }
}

// MARK: - ChunkPacket

/// Parsed CHUNK packet.
///
/// Wire layout (64-bit / primary):
///   "CHUNK" (5B) | frame_id (UInt64 LE) | chunk_index (UInt64 LE) | payload (≤1200B)
///   Header: 21 bytes
///
/// Wire layout (32-bit / fallback):
///   "CHUNK" (5B) | frame_id (UInt32 LE) | chunk_index (UInt32 LE) | payload (≤1200B)
///   Header: 13 bytes
public struct ChunkPacket {
    public let frameID:     UInt64
    public let chunkIndex:  UInt64
    public let payload:     Data

    public static func parse(from data: Data) -> ChunkPacket? {
        let marker = VideoProtocol.chunkMarker
        guard data.starts(with: marker) else { return nil }

        let header64Size = marker.count + 16   // 21 bytes
        let header32Size = marker.count + 8    // 13 bytes

        if data.count >= header64Size {        // Try 64-bit first (primary)
            let parsed: ChunkPacket? = data.dropFirst(marker.count).withUnsafeBytes { ptr in
                guard ptr.count >= 16 else { return nil }
                let frameID    = UInt64(littleEndian: ptr.loadUnaligned(fromByteOffset: 0, as: UInt64.self))
                let chunkIndex = UInt64(littleEndian: ptr.loadUnaligned(fromByteOffset: 8, as: UInt64.self))
                let payload    = data.suffix(from: header64Size)
                return ChunkPacket(frameID: frameID, chunkIndex: chunkIndex, payload: payload)
            }
            if let parsed { return parsed }
        }

        if data.count >= header32Size {        // 32-bit fallback
            return data.dropFirst(marker.count).withUnsafeBytes { ptr in
                guard ptr.count >= 8 else { return nil }
                let frameID    = UInt64(UInt32(littleEndian: ptr.loadUnaligned(fromByteOffset: 0, as: UInt32.self)))
                let chunkIndex = UInt64(UInt32(littleEndian: ptr.loadUnaligned(fromByteOffset: 4, as: UInt32.self)))
                let payload    = data.suffix(from: header32Size)
                return ChunkPacket(frameID: frameID, chunkIndex: chunkIndex, payload: payload)
            }
        }

        return nil
    }
}
